server/src/main/java/org/elasticsearch/action/ingest/WriteableIngestDocument.java

```diff
@@ -83,9 +83,14 @@ final class WriteableIngestDocument implements Writeable, ToXContentFragment {
         PARSER.declareObject(constructorArg(), INGEST_DOC_PARSER, new ParseField(DOC_FIELD));
     }
 
+    /**
+     * Builds a writeable ingest document that wraps a copy of the passed-in, non-null ingest document.
+     *
+     * @throws IllegalArgumentException if the passed-in ingest document references itself
+     */
     WriteableIngestDocument(IngestDocument ingestDocument) {
         assert ingestDocument != null;
-        this.ingestDocument = ingestDocument;
+        this.ingestDocument = new IngestDocument(ingestDocument); // internal defensive copy
```

Comment:
I'm not concerned about simulate performance, but this does mean we'll make an extra deepCopy and do an extra reference check every time the xcontent parser runs (line 60 above). I don't think that's performance-sensitive though right?

Reply:
Joe pointed out offline that WriteableIngestDocument is only used for the simulate API, so my performance concern here is irrelevant.